### PR TITLE
kd-tree kNN search

### PR DESCRIPTION
**Pull request by averbukh**
# Description
Extension of kd-tree kNN

Changes proposed in this pull request:
- optional results: indices and distances
- voting weights: uniform (old behavior) and distances

### CODE REVIEW

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h
Comment (Alexsandruss) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h:183]: may be, interface3?

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h
Comment (Alexsandruss) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h:65]: computeClassLabels might be added to skip class assignments

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h
Comment (Alexsandruss) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h:123]: DAAL-CLASS-ALGORITHMS__KDTREE_KNN_CLASSIFICATION__PREDICTION__RESULT

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.cpp
Comment (Alexsandruss) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.cpp:78]: Why checks are commented?

Comment (averbukh) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.cpp:78]: I skipped it for prototype, to prepare PR as soon as possible to unblock Python and testing activities, it's part of productization (to be part of next commits into this PR)

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h
Comment (averbukh) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h]: Fixed, thank you

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h
Comment (averbukh) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h:65]: Not sure that are you talking about. computeClassLabels is in resultToEvaluate, not resultToCompute

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h
Comment (averbukh) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_types.h:183]: No, interfaceN are not consistent for any alg anyway, because we add new interfaceN for classes actually changed, so there are no reasons to skip some numbers in interfaceN names.

**Comment by averbukh** [comment]
> I haven't question about implementation, but you should verify that interfaces correctly changed. At least there was a problem of building daal4py connected to interface1::Result.

Is it still problem for daal4py? If yes, we need to discuss it to find best way to fix it. I really think that skipping of some N in interfaceN names is bad idea, but if it affects daal4py, I would like to discuss it with @PetrovKP  and @Alexander-Makaryev.

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict.h
Comment (ShvetsKS) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict.h:334]: why we have parameter as a value here?
seems `parameter()` method should be added.

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h
Comment (ShvetsKS) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h:234]: similar comment for `predict::Batch` class

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h
Comment (ShvetsKS) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h:374]: parameter()

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_model_impl.cpp
Comment (ShvetsKS) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_model_impl.cpp:86]: what is the reason to delete supper check?

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_input.cpp
Comment (ShvetsKS) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_input.cpp:46]: is it possible to have some common function for similar check in base class https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/classifier/classifier_train.cpp#L92 ?

**Review by ShvetsKS — CHANGES_REQUESTED** [review]
changes related to explicit interfaceN providing are required
and if it's possible `parameter()` method should be added

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict.h
Comment (averbukh) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict.h:334]: Sure. I plan to add it with another modifications for test system by another PR

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_model_impl.cpp
Comment (averbukh) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_model_impl.cpp:86]: Because here, we need relaxed check relative to inherited

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_input.cpp
Comment (averbukh) [reply cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_input.cpp:46]: It's possible, but no reason to do it now due to new interfaces.

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h
Comment (averbukh) [reply cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_training_batch.h:234]: Great catch! Thank you.

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/maverbuk-knn

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/maverbuk-knn

**Review by ShvetsKS — APPROVED** [review]
LGTM

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h
Comment (Alexsandruss) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h:186]:
```suggestion
    Parameter(size_t nClasses = 2, size_t nNeighbors = 1, int randomSeed = 777, DataUseInModel dataUse = doNotUse, DAAL_UINT64 resToCompute = 0, DAAL_UINT64 resToEvaluate = 0,
```

File: cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h
Comment (Alexsandruss) [inline cpp/daal/include/algorithms/k_nearest_neighbors/kdtree_knn_classification_model.h:195]:
```suggestion
    {
        this->resultsToEvaluate = resToEvaluate;
    }
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
To correctly handle resultsToEvaluate, following changes are needed.
Also, check `DAAL_CHECK_EX(resultsToEvaluate != 0, services::ErrorIncorrectParameter, services::ParameterName, resultsToEvaluateStr())` (line 43) should be deleted from cpp/daal/src/algorithms/classifier/classifier_train.cpp

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/maverbuk-knn

**Comment by averbukh** [comment]
/intelci: restart
set infra_branch=dev/maverbuk-knn

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/maverbuk-knn

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/ntimakin-kdtree_knn_tests

**Comment by Alexsandruss** [comment]
@averbukh, please, add `none = 0x00000000ULL` to ResultToComputeId of classifier in cpp/daal/include/algorithms/classifier/classifier_model.h [line 47] to support label optionality in kNN search python wrapper.

**Comment by averbukh** [comment]
> @averbukh, please, add `none = 0x00000000ULL` to ResultToComputeId of classifier in cpp/daal/include/algorithms/classifier/classifier_model.h [line 47] to support label optionality in kNN search python wrapper.

As discussed offline, it isn't necessary for kd-tree kNN, because it's part of PR for brute-force kNN.

**Review by Alexsandruss — APPROVED** [review]

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.h
Comment (Alexsandruss) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.h:24]:
```suggestion
#ifndef __KDTREE_KNN_CLASSIFICATION_PREDICTION_RESULT_
```

File: cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.h
Comment (Alexsandruss) [inline cpp/daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_predict_result.h:25]:
```suggestion
#ifndef __KDTREE_KNN_CLASSIFICATION_PREDICTION_RESULT_
#define __KDTREE_KNN_CLASSIFICATION_PREDICTION_RESULT_
```

**Comment by averbukh** [comment]
/intelci: run
set infra_branch=dev/ntimakin-kdtree_knn_tests

**Comment by averbukh** [comment]
Ci is green except of compatibility testing, which is expected and to be fixed after merge of corresponding testing from merge request to appropriate branch.

**Pull request merged**
